webhooks/virtualmachinegrouppublishrequest/mutation/virtualmachinegrouppublishrequest_mutator.go

```diff
@@ -0,0 +1,152 @@
+	return mutator{
+		client:    client,
+		converter: runtime.DefaultUnstructuredConverter,
+	}
+}
+
+type mutator struct {
+	client    ctrlclient.Client
+	converter runtime.UnstructuredConverter
+}
+
+func (m mutator) For() schema.GroupVersionKind {
+	return vmopv1.GroupVersion.WithKind(reflect.TypeOf(vmopv1.VirtualMachineGroupPublishRequest{}).Name())
+}
+
+func (m mutator) Mutate(ctx *pkgctx.WebhookRequestContext) admission.Response {
+	// this check is unnecessary but serves as a saft guard
+	if ctx.Op != admissionv1.Create {
+		return admission.Allowed("")
+	}
+
+	return m.mutateCrate(ctx)
+}
+
+// mutateCrate sets spec.source to the VM group publish request's name if it is omitted,
+// spec.target to a default content library if one exists, and spec.virtualMachines to
+// all the VMs under source VM group.
+func (m mutator) mutateCrate(ctx *pkgctx.WebhookRequestContext) admission.Response {
+	groupPubReq, err := m.vmGroupPublishRequestFromUnstructured(ctx.Obj)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	if groupPubReq.Spec.Source == "" {
+		groupPubReq.Spec.Source = groupPubReq.Name
+	}
+
+	if groupPubReq.Spec.Target == "" {
+		defaultCL, err := common.RetrieveDefaultImagePublishContentLibrary(ctx, m.client, groupPubReq.Namespace)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return admission.Errored(http.StatusNotFound, err)
+			}
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+		groupPubReq.Spec.Target = defaultCL.Name
+	}
+
+	if len(groupPubReq.Spec.VirtualMachines) == 0 {
+		vms, err := m.retrieveVMGroupMembers(ctx, ctrlclient.ObjectKey{Namespace: groupPubReq.Namespace, Name: groupPubReq.Spec.Source})
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return admission.Errored(http.StatusNotFound, err)
+			}
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+		groupPubReq.Spec.VirtualMachines = vms
+	}
+	// ensure vms are unique and sorted
+	groupPubReq.Spec.VirtualMachines = sets.List(sets.New(groupPubReq.Spec.VirtualMachines...))
+
+	rawGroupPubReq, err := json.Marshal(groupPubReq)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+	return admission.PatchResponseFromRaw(ctx.RawObj, rawGroupPubReq)
+}
+
+// retrieveVMGroupMembers retrieves all the VMs under a VM group recursively.
+func (m mutator) retrieveVMGroupMembers(ctx *pkgctx.WebhookRequestContext, vmGroupKey ctrlclient.ObjectKey) ([]string, error) {
+	var vms []string
+
+	vmGroup := &vmopv1.VirtualMachineGroup{}
+	if err := m.client.Get(ctx, vmGroupKey, vmGroup); err != nil {
+		return nil, err
+	}
+
+	for _, member := range vmGroup.Status.Members {
+		switch member.Kind {
+		case "VirtualMachine":
+			vms = append(vms, member.Name)
+		case "VirtualMachineGroup":
+			groupMembers, err := m.retrieveVMGroupMembers(ctx, ctrlclient.ObjectKey{Namespace: vmGroupKey.Namespace, Name: member.Name})
```

Comment:
Can you create an issue to handle the possible loop / recursive depth? (The same issue exists in the Group controller)

Reply:
ack, linked to vm group publish epic.